Firestore/core/src/util/defer.h

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_UTIL_DEFER_H_
+#define FIRESTORE_CORE_SRC_UTIL_DEFER_H_
+
+#include <utility>
+
+namespace firebase {
+namespace firestore {
+namespace util {
+
+template <typename Action>
+class Deferred;
+
+/**
+ * Creates a deferred action that will be destroyed at the close of the lexical
+ * scope containing the result of this function call. This is useful for
+ * performing ad-hoc RAII-style actions, without having to create the wrapper
+ * class. For example:
+ *
+ *     FILE* file = fopen(filename, "rb");
+ *     auto cleanup = defer([&] {
+ *       if (file) {
+ *         fclose(file);
+ *       }
+ *     });
+ *
+ * @param action a callable object; usually a lambda. Even if exceptions are
+ *     enabled, when `action` is invoked it must not throw. This is similar to
+ *     the restriction that exists on destructors generally.
+ */
+template <typename Action>
+Deferred<Action> defer(Action&& action) {
+  return Deferred<Action>(std::forward<Action>(action));
+}
+
+/**
+ * Storage for a deferred action. The `action` is invoked during the destructor
+ * of the `Deferred`.
+ */
+template <typename Action>
+class Deferred {
+ public:
+  explicit Deferred(Action&& action) : action_(std::move(action)) {
+  }
+
```

Comment:
Should this class forbid copying (and probably moving as well)?

Reply:
In order to return this by value the type must at least be moveable. Only once we get to C++17 with guaranteed elision can we return a non-copyable, non-moveable object by value. I've made it move-only.

The other alternative is to make this class concrete, taking a `std::function<void()>` as an argument. Then the constructor can be invoked directly and the class could be made non-copyable and non-moveable (like `std::lock_guard`). Usage would look like this:

```c++
FILE* file = fopen(filename, "rb");
Defer cleanup([&] {
  if (file) {
    fclose(file);
  }
}
```

Unfortunately, the major downside of that approach is that `std::function` must perform a heap allocation whereas the inferred implementation defined type of `Action` here likely does not.